README.txt
===[CONTENTS]==================================================================

1 - ABOUT
2 - LICENSE/DISCLAIMER
3 - NON-EXHAUSTIVE FEATURE LIST
4 - INSTALLATION AND SET UP

===[1 - ABOUT]=================================================================

FOSSTube v1.6 (December 2022)
Copyright (C) 2020-2022 Norbert de Jonge <nlmdejonge@gmail.com>

A free and open-source video sharing content management system.

The FOSSTube website can be found at [ https://github.com/EndeavourAccuracy/FOSSTube ].

===[2 - LICENSE/DISCLAIMER]====================================================

This software is provided 'as-is', without any express or implied
warranty.  In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgment in the product documentation would be
   appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.

--------------------
The above (FSF and OSI approved) zlib license is only for FOSSTube itself.
Distro, website and platform icons are property of their respective owners.
Bootstrap 3.3.7: MIT
HTML5 Shiv 3.7.3: MIT/GPL2
IE10 viewport hack: MIT
jQuery 3.3.1: MIT
noUiSlider 14.0.3: MIT
Respond.js 1.4.2: MIT
Swift Mailer 4.3.0: LGPL3
wNumb 1.2.0: MIT
PHP QR Code 1.1.4: LGPL3
[Used only for spherical videos:] Video.js (VR) 7.13.3 (1.8.0): Apache 2.0

===[3 - NON-EXHAUSTIVE FEATURE LIST]===========================================

* Users can publish videos (accepts all video formats, HTML5 video playback).
* Users can publish text, such as blog posts, with (previewable) BBCode markup.
* Forum, with boards and topics.
* Responsive (mobile/tablet friendly) design.
* Comments, including conversation threading.
* Social media integration (share thumbnails; Open Graph and Twitter Cards).
* User profile pages, with avatars and custom text.
* Public and private folders for bookmarking and creating collections.
* Both automatically extracted and custom thumbnails.
* Channel subscriptions.
* Notifications, for new content, comments/replies, admin messages.
* Content titles, descriptions, tags, categories, restrictions, languages.
* Per-content comment settings (on, off, only approved).
* Content and comments can be liked (thumbs up).
* Real-time switching between sizes (360p, 720p, 1080p).
* User roles (owners, admins, mods) and optional per-user privileges.
* Per-user monetization options (processors, cryptocurrencies), including a modal dialog.
* Sitewide 4-byte UTF-8 Unicode support (inc. emoji).
* Search, with autocomplete and hits preview.
* Advanced search page, with optional user filter, custom ordering, etc.
* Extensive content reporting options, a moderation queue, and banning options.
* Content pagination.
* Server-side transcoding queue.
* Subtitle support (via WebVTT).
* Sitewide and per-user RSS feeds.
* Multiple file upload support, with progress bars.
* Extensive admin panel, with user messaging, video purging, Tor blocking, etc.
* On-hover video previews.
* Sitewide and per-user referrer saving.
* Users can put up videos for adoption.
* Optional iframe embedding of videos.
* Hotlink protection.
* HTML5 valid pages (Nu Html Checker).
* Loop functionality, including customized links.
* Home page filters: views threshold, (not) safe for work.
* Extracts and displays video durations and average FPSs; counts views.
* Publishers can love/pin comments, (un)mute users.
* Lots of security measures, such as hashed passwords created on-site, sanitization, tokens, CAPTCHAs, logging of failed login attempts, and verification codes, to protect against CSRF, SQL injection, XSS, and brute-force attacks.
* User content is sortable by date, views, likes, comments, and durations.
* Auto-generation of related content.
* Live, development and maintenance modes.
* Feedback forms (general, when deleting accounts).
* IPv6 support.
* Clickable time stamps.
* Optional night mode.
* Moderators can move forum topics and can change (N)SFW statuses.
* Sticky (hideable) floatable videos when scrolling down long pages.
* Forgot username/password functionality.
* Users can change their email addresses, passwords, and usernames.
* Video playback speed control.
* Users can browse their comment history and subscriptions.
* User preferences: show NSFW content, home container width, thumbnail size, fewer notifications about certain users.
* Users can send and receive information requests (email address).
* Admins can change forum topic titles, and lock forum topics.
* Customizable 404 Not Found page.
* Disallows users from adding duplicate videos to the same account (MD5), and the video overview lists duplicate videos on other accounts (again, MD5).
* Polls can be attached to forum topics.
* Support for spherical videos (with Video.js).
* Highlighting of unread topics and comments/replies.
* Twitter-like microblogging on user pages, including likes, reblogs, following and timeline, hashtags and hashflags, status pages, and explore (search).
* Featured content.
* Trending page, based on daily top 10 logs.
* Basic per-user statistics.
* Optional hat-note.
* Optional DMCA page.

===[4 - INSTALLATION AND SET UP]===============================================

--------------------
Before you get started
--------------------
FOSSTube is fairly easy to install and set up, but it requires at least some knowledge of GNU/Linux. If you have no experience with GNU/Linux, I urge you to ask a system administrator to get FOSSTube up and running for you.

FOSSTube is an asynchronous web application that heavily uses Ajax, which requires visitors to have JavaScript enabled. FOSSTube is NOT a decentralized, federated or peer-to-peer application. FOSSTube transcodes to MP4 H.264. Account creation requires a valid email address.

An example website that uses FOSSTube:
https://www.freespeechtube.org/
(FOSSTube is NOT associated with FreeSpeechTube.)

--------------------
Requirements
--------------------
* Apache HTTP Server 2
* MySQL/MariaDB
* PHP 8 (or 7)
* FFmpeg (ffmpeg + ffprobe)
* The ability to set up and run cronjobs.
* An SMTP account, preferably with an accurate SPF-record.
* A domain name.
* A proper server. This is subjective, but I suggest using 4+ CPUs, 8G+ RAM, 100G+ SSD, and being allowed to have 2TB+ monthly traffic.

Unless you feel comfortable using it, do NOT use DirectAdmin.

--------------------
Settings and customization
--------------------
The assumption is that FOSSTube will be installed in the root of your domain. Running FOSSTube from a subdirectory is completely untested. You may attempt doing so, e.g. by including a <base> in HTMLStart(), but if you do, you are on your own.

You must have a non-public directory (such as private/) on the same level as your website directory (such as public_html/ or www/). Virtually all accounts have this nowadays. If your account does not, (ask your webmaster to) create one.

First, check if any addendums have been issued for this text:
https://github.com/EndeavourAccuracy/FOSSTube/issues

Add
character-set-server=utf8mb4
to your MySQL configuration.
$ sudo vim /etc/mysql/mysql.conf.d/mysqld.cnf
At the end of the [mysqld] section, add:
character-set-server=utf8mb4
$ sudo service mysql restart

Edit FOSSTube.sql to change all instances of "YOURPASS", and optionally also change the name of forum board "FOSSTube".
Then use FOSSTube.sql to create all MySQL tables:
$ sudo mysql < FOSSTube.sql

Enable various Apache and PHP extensions:
$ sudo a2enmod rewrite
$ sudo a2enmod headers
$ sudo apt install php-mysql
$ sudo apt install php-mbstring
$ sudo apt install php-gd
$ sudo apt install php-cli
$ sudo service apache2 restart

Modify various PHP settings:
$ sudo vim /etc/php/8.x/apache2/php.ini
file_uploads = On
max_execution_time = 1500
max_input_time = 60
max_input_vars = 1000
memory_limit = 500M
post_max_size = 500M
session.gc_maxlifetime = 1440
upload_max_filesize = 500M
zlib.output_compression = Off
$ sudo service apache2 restart

Modify all "CHANGE" in private/fst_db.php.
(Matching your CREATE USER line of FOSSTube.sql.)

Modify all settings in private/fst_settings.php.
If, for security purposes, you renamed the directories "swift_random" and "phpqrcode_random", also change it in the above file.

Modify public_html/.htaccess.
At the very least the two instances of "yourdomain.com".
Also, if you put $GLOBALS['live'] on FALSE in fst_settings.php, then add "#" in front of all lines under "# DO NOT USE ON LOCALHOST".

Edit public_html/templates/default.htm and replace "YOURIMAGE" with a Base64 image. You can do this by using a search engine to find a 'base64 image encoder' and then create something like this:
<img src="data:image/png;base64,..." alt="yourdomain">

Modify HTML text:
* public_html/about/about.html
* public_html/terms/terms.html
* public_html/privacy/privacy.html
* public_html/mod/guidelines.html
* public_html/404.html
* public_html/patronage/patronage.html
If you comment out the DMCA policy in fst_base.php, then also modify:
* public_html/DMCA/DMCA.html

Modify images:
* All images in the directory public_html/images/favicons/
* public_html/images/yourdomain_208x30.png
* public_html/images/avatar_small.png
* public_html/images/avatar_large.png
* public_html/images/back_01.jpg
* public_html/images/back_02.jpg

Create two cronjobs that run every minute (all *):
php (path)/private/fst_encode.php >/dev/null 2>&1
php (path)/private/fst_actions.php >/dev/null 2>&1
Obviously, replace (path) with the full path.

Upload all private/ and public_html/ files.

Create an administrator account (you specified their usernames in fst_settings.php), and login. Then click "Admin" to make sure no red error messages show up.

If uploading fails, try editing public_html/upload/upload.php and change
$GLOBALS['ffprobe'] = substr (shell_exec ('which ffprobe'), 0, -1);
to
$GLOBALS['ffprobe'] = '/bin/ffprobe';
(Or wherever ffprobe resides.)

Questions? Contact nlmdejonge@gmail.com.


ChangeLog.txt
v1.6 (December 1, 2022) - Norbert de Jonge
--------------------
+ Optional hat-note.
+ Optional DMCA page.
* Minor improvements and fixes.

v1.5 (February 11, 2022) - Norbert de Jonge
--------------------
+ Basic per-user statistics.
* Software renamed from FSTube to FOSSTube, to fully disassociate from FreeSpeechTube.
* Minor improvements.

v1.4 (December 12, 2021) - Norbert de Jonge
--------------------
+ Twitter-like microblogging on user pages, including likes, reblogs, following and timeline, hashtags and hashflags, status pages, and explore (search).
+ Featured content.
+ Trending page, based on daily top 10 logs.
* Fixed a minor poll voting bug.

v1.3 (September 9, 2021) - Norbert de Jonge
--------------------
+ Advanced search page.
+ Additional user preference: fewer notifications about certain users.

v1.2 (August 8, 2021) - Norbert de Jonge
--------------------
+ Polls can now be attached to forum topics.
+ Support for spherical videos (with Video.js).
+ Highlighting of unread topics and comments/replies.
+ Separate unmute page.
+ Admins can lock forum topics.

v1.1 (March 19, 2021) - Norbert de Jonge
--------------------
+ Video playback speed control.
+ Users can browse their comment history and subscriptions.
+ User preferences: home container width, thumbnail size.
+ Users can send and receive information requests (email address).
+ Admins can change forum topic titles.
+ Customizable 404 Not Found page.
+ Disallows users from adding duplicate videos to the same account (MD5), and the video overview lists duplicate videos on other accounts (again, MD5).
* In addition to processors, monetization now includes cryptocurrencies.
* Expanded the admin panel.

v1.0 (February 4, 2020) - Norbert de Jonge
--------------------
+ Initial release.


public_html/js/fst.js
/* SPDX-License-Identifier: Zlib */
/* FOSSTube v1.6 (December 2022)
 * Copyright (C) 2020-2022 Norbert de Jonge <nlmdejonge@gmail.com>
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
 */

function BackSignUp () {
	$.ajax({
		type: "POST",
		url: "/signup/step_back.php",
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			if (result == 1)
			{
				location.reload();
			} else {
				alert (error);
			}
		},
		error: function() {
			alert ("Error calling step_back.php.");
		}
	});
}

function BackForgot () {
	$.ajax({
		type: "POST",
		url: "/forgot/step_back.php",
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			if (result == 1)
			{
				location.reload();
			} else {
				alert (error);
			}
		},
		error: function() {
			alert ("Error calling step_back.php.");
		}
	});
}

function BackAccount () {
	$.ajax({
		type: "POST",
		url: "/account/step_back.php",
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			if (result == 1)
			{
				location.reload();
			} else {
				alert (error);
			}
		},
		error: function() {
			alert ("Error calling step_back.php.");
		}
	});
}

function Poll () {
	var poll = $("#poll-div").data("poll");
	$.ajax({
		type: "POST",
		url: "/v/poll.php",
		data: ({
			poll : poll
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			if (result == 1)
			{
				$("#poll-div").html(html);
			} else {
				$("#poll-error").html(error);
			}
		},
		error: function() {
			$("#poll-error").html("Error calling poll.php.");
		}
	});
}

function Comments (comment) {
	var code = $("#comments").data("code");
	$.ajax({
		type: "POST",
		url: "/v/comments.php",
		data: ({
			code : code,
			comment : comment
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			if (result == 1)
			{
				$("#comments").html(html);
			} else {
				$("#comment-error").html(error);
			}
		},
		error: function() {
			$("#comment-error").html("Error calling comments.php.");
		}
	});
}

function CommentToggle (toggle, comment_id, csrf_token) {
	$.ajax({
		type: "POST",
		url: "/v/toggle.php",
		data: ({
			toggle : toggle,
			comment_id : comment_id,
			csrf_token : csrf_token
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var state = data["state"];
			if (result == 1)
			{
				$("#" + toggle + "-" + comment_id).html("<a name=\"" + toggle + "\" href=\"javascript:;\" style=\"display:inline-block;\"><img src=\"/images/" + toggle + "_" + state + ".png\" alt=\"" + toggle + " " + state + "\"></a>");
				if ((toggle == "pinned") || (toggle == "hidden") || (toggle == "muted")) { Comments (0); }
				if (toggle == "approved")
				{
					if (state == "on") { content = ""; } else { content = "As the content publisher, you can press the gray check mark to approve this comment."; }
					$("#approved-hint-" + comment_id).html(content);
				}
			} else {
				alert (error);
			}
		},
		error: function() {
			alert ("Error calling toggle.php.");
		}
	});
}

function VideosJS (divid, section, order, offset,
	user, searcht, searcha, filters) {
/***
	var filtersa = [];
	$.each(filters, function(key,value) {
		var str = key + ":" + value;
		filtersa.push(str);
	});
***/
	$.ajax({
		type: "POST",
		url: "/videos.php",
		data: ({
			section : section,
			order : order,
			offset : offset,
			user : user,
			searcht : searcht,
			searcha : searcha,
			filters : filters
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			if (result == 1)
			{
				$("#" + divid).html(html);
			} else {
				alert (error);
			}
		},
		error: function(xhr, status, error) {
			console.log(xhr.responseText);
			console.log(status);
			console.log(error);
			alert ("Error calling videos.php.");
		}
	});
}

function UpdateSearchList (keycode, search_query) {
	if ((keycode == 37) || (keycode == 38) || (keycode == 39) || (keycode == 40))
		{ return false; }

	if (search_query.length < 3)
	{
		$("#matches_title").html("");
		$("#matches_any").html("");
		$("#search_list").html("");
		$("#search_query").css("background-color","#f8f8ff");
		return false;
	}

	$.ajax({
		type: "POST",
		url: "/searchlist.php",
		data: ({
			search_query : search_query
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			var matches_title = data["matches_title"];
			var matches_any = data["matches_any"];
			if (result == 1)
			{
				$("#matches_title").html(matches_title);
				$("#matches_any").html("(+" + (matches_any - matches_title) + ")");
				$("#search_list").html(html);
				if ((matches_title == 0) && (matches_any == 0))
					{ var bgcolor = "#fcc"; } else { var bgcolor = "#f8f8ff"; }
				$("#search_query").css("background-color",bgcolor);
			} else {
				alert (error);
			}
		},
		error: function() {
			alert ("Error calling searchlist.php.");
		}
	});
}

function PadLeft (nr) {
	if (nr < 10) { nr = "0" + nr; }
	return (nr);
}

function SecToTime (sec) {
	var seconds = Math.floor(sec % 60);
	var minutes = Math.floor((sec / 60) % 60);
	var hours = Math.floor((sec / (60 * 60)) % 60);

	return PadLeft(hours) + ":" + PadLeft(minutes) + ":" + PadLeft(seconds);
}

function Size (size, projection, src) {
	if (projection == 0)
	{
		var fvideo = $("#video")[0];
		var ttime = fvideo.currentTime;
		fvideo.src = src;
		fvideo.currentTime = ttime;
		fvideo.play();
	} else {
		var fvideo = videojs("video");
		var ttime = fvideo.currentTime();
		fvideo.src({ src: src });
		fvideo.currentTime(ttime);
		fvideo.play();
	}
	switch (size) {
		case "360":
			if($("#q360").length > 0) { $("#q360").addClass("activep"); }
			if($("#q720").length > 0) { $("#q720").removeClass("activep"); }
			if($("#q1080").length > 0) { $("#q1080").removeClass("activep"); }
			break;
		case "720":
			if($("#q360").length > 0) { $("#q360").removeClass("activep"); }
			if($("#q720").length > 0) { $("#q720").addClass("activep"); }
			if($("#q1080").length > 0) { $("#q1080").removeClass("activep"); }
			break;
		case "1080":
			if($("#q360").length > 0) { $("#q360").removeClass("activep"); }
			if($("#q720").length > 0) { $("#q720").removeClass("activep"); }
			if($("#q1080").length > 0) { $("#q1080").addClass("activep"); }
			break;
	}
}

function Boards () {
	$.ajax({
		type: "POST",
		url: "/forum/boards.php",
		data: ({
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			if (result == 1)
			{
				$("#boards").html(html);
			} else {
				$("#boards-error").html(error);
			}
		},
		error: function() {
			$("#boards-error").html("Error calling boards.php.");
		}
	});
}

function Topics (offset) {
	var board = $("#topics").data("board");
	$.ajax({
		type: "POST",
		url: "/forum/topics.php",
		data: ({
			board : board,
			offset : offset
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			if (result == 1)
			{
				$("#topics").html(html);
			} else {
				$("#topics-error").html(error);
			}
		},
		error: function() {
			$("#topics-error").html("Error calling topics.php.");
		}
	});
}

function TrendingDates (curdate, actyear) {
	$.ajax({
		type: "POST",
		url: "/trending/dates.php",
		data: ({
			curdate : curdate,
			actyear : actyear
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			var actdate = data["actdate"];
			if (result == 1)
			{
				$("#trending-date").html(html);
				Trending (actdate);
			} else {
				$("#trending-error").html(error);
			}
		},
		error: function() {
			$("#trending-error").html("Error calling dates.php.");
		}
	});
}

function Trending (date) {
	$.ajax({
		type: "POST",
		url: "/trending/trending.php",
		data: ({
			date : date
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			if (result == 1)
			{
				$("#trending").html(html);
			} else {
				$("#trending-error").html(error);
			}
		},
		error: function() {
			$("#trending-error").html("Error calling trending.php.");
		}
	});
}

function MicroBlogPosts (user, mboffset) {
	$.ajax({
		type: "POST",
		url: "/timeline/mbposts.php",
		data: ({
			user : user,
			mboffset : mboffset
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var html = data["html"];
			var posts = data["posts"];
			if (result == 1)
			{
				if (mboffset == 0)
				{
					$("#microblogposts").html(html);
				} else {
					$("#microblogposts").append(html);
				}
				if (posts == 0)
				{
					$("#microblogposts-more").hide();
				} else {
					$("#microblogposts-more").show();
				}
			} else {
				$("#microblogposts-error").html(error);
			}
		},
		error: function() {
			$("#microblogposts-error").html("Error calling mbposts.php.");
		}
	});
}

function HidePost (post_id, csrf_token) {
	$.ajax({
		type: "POST",
		url: "/user/mbpost_hide.php",
		data: ({
			post_id : post_id,
			csrf_token : csrf_token
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			if (result == 1)
			{
				var post = "#post-" + post_id;
				$(post).html("<i>Removed.</i>");
				$(post).delay(1000).fadeOut(500);
			} else {
				alert (error);
			}
		},
		error: function() {
			alert ("Error calling mbpost_hide.php.");
		}
	});
}

function LikePost (post_id, csrf_token) {
	$.ajax({
		type: "POST",
		url: "/user/mbpost_like.php",
		data: ({
			post_id : post_id,
			csrf_token : csrf_token
		}),
		dataType: "json",
		success: function(data) {
			var result = data["result"];
			var error = data["error"];
			var likes = data["likes"];
			if (result == 1)
			{
				$("#liked-" + post_id).html("<img src=\"/images/liked_on.png\" alt=\"liked on\">");
				$("#likes-" + post_id).html(likes);
			} else {
				alert (error);
			}
		},
		error: function() {
			alert ("Error calling mbpost_like.php.");
		}
	});
}

$(document).ready(function(){
	$("[data-name=\"hover\"]").mouseenter(function(){
		var active = $(this).data("active");
		var preview = $(this).data("preview");
		if ((active == "thumb") && (preview != ""))
		{
			var title = $(this).data("title");
			title = title.replace(/\"/g, "&quot;");
			style = $(this).find("img").attr("style");
			if (style == "undefined")
				{ style = ""; }
					else { style = " style=\"" + style + "\""; }
			var video = "<video src=\"" + preview + "\" alt=\"" +
				title + "\" class=\"thumb-or-preview\"" + style + " autoplay loop>";
			$(this).html(video);
			$(this).data("active","video");
		}
	});
	$("[data-name=\"hover\"]").mouseleave(function(){
		var active = $(this).data("active");
		if (active == "video")
		{
			var thumb = $(this).data("thumb");
			var title = $(this).data("title");
			title = title.replace(/\"/g, "&quot;");
			style = $(this).find("video").attr("style");
			if (style == "undefined")
				{ style = ""; }
					else { style = " style=\"" + style + "\""; }
			var image = "<img src=\"" + thumb + "\" alt=\"" +
				title + "\" class=\"thumb-or-preview\"" + style + ">";
			$(this).html(image);
			$(this).data("active","thumb");
		}
	});

	$(window).scroll(function(){
		if ($("#video-div").length > 0) {
			var bottom = $("#video-div").offset().top + $("#video-div").height();
			var hover = $("#video-div").data("hover");
			var down = $(window).scrollTop();
			if (down > bottom) {
				if (hover == "no") {
					$("#video-div").css("height",$("#video").height());
					$("#video-cont-div").addClass("hover");
					$("#video-cont-span").css("display","block");
					$("#video-div").data("hover","yes");
				}
			} else {
				$("#video-div").css("height","auto");
				$("#video-cont-div").removeClass("hover");
				$("#video-cont-span").css("display","none");
				$("#video-div").data("hover","no");
				/***/
				$("#video").css("display","block");
				$("#video-cont-a").text("hide");
				$("#video").data("hidden","no");
			}
		}
	});

	$("#switch").click(function(){
		var theme = $("#theme").data("theme");

		$.ajax({
			type: "POST",
			url: "/switch.php",
			data: ({
				theme : theme
			}),
			dataType: "json",
			success: function(data) {
				var result = data["result"];
				var error = data["error"];
				var theme = data["theme"];
				if (result == 1)
				{
					if (theme == "day")
					{
						$("#theme").data("theme","day");
						$("#theme").attr("href","/css/fst_day.css?v=24");
						$("#switch-img").attr("alt","night");
						$("#switch-img").attr("src","/images/theme/night.png");
					} else {
						$("#theme").data("theme","night");
						$("#theme").attr("href","/css/fst_night.css?v=24");
						$("#switch-img").attr("alt","day");
						$("#switch-img").attr("src","/images/theme/day.png");
					}
				} else {
					alert (error);
				}
			},
			error: function() {
				alert ("Error calling switch.php.");
			}
		});
	});

	var MaxHeight = 60;
	$(".limit-height").each(function(){
		var scrollh = this.scrollHeight;
		var div = $(this);
		if (scrollh > MaxHeight) {
			div.css("height",MaxHeight + "px");
			div.after("<a href=\"javascript:;\">Show more</a><br>");
			var link = div.next();
			link.click(function(e){
				e.stopPropagation();
				if (link.text() != "Show less") {
					link.text("Show less");
					div.animate({"height": scrollh});
				} else {
					link.text("Show more");
					div.animate({"height": MaxHeight + "px"});
				}
			});
		}
	});

	if (window.location.hash)
	{
		var hash = window.location.hash;
		$('html, body').animate({
			scrollTop: $(hash).offset().top
		}, 500);
	}
});
